solr/core/src/java/org/apache/solr/cloud/api/collections/CreateCollectionCmd.java

```diff
@@ -190,10 +190,12 @@ public void call(ClusterState clusterState, ZkNodeProps message, NamedList resul
       try {
         replicaPositions = buildReplicaPositions(ocmh.cloudManager, clusterState, clusterState.getCollection(collectionName), message, shardNames, sessionWrapper);
       } catch (Assign.AssignmentException e) {
-        ZkNodeProps deleteMessage = new ZkNodeProps("name", collectionName);
-        new DeleteCollectionCmd(ocmh).call(clusterState, deleteMessage, results);
+        deleteCollection(clusterState, results, collectionName);
         // unwrap the exception
         throw new SolrException(ErrorCode.SERVER_ERROR, e.getMessage(), e.getCause());
+      } catch (SolrException e) {
```

Comment:
So one question I have is why is the error coming back from `buildReplicaPositions` not an `Assign.AssignmentException`? Is it because it is wrapped in a `SolrException` from the remote node?

Reply:
`buildReplicaPositions` checks the max shard number [itself](https://github.com/apache/lucene-solr/blob/releases/lucene-solr/8.4.1/solr/core/src/java/org/apache/solr/cloud/api/collections/CreateCollectionCmd.java#L389-L390) and throws a new `SQLException`. It also calls `Assign.*` methods, those might throw the `Assign.AssignmentException` exceptions.